### PR TITLE
Enable snapshot/restore functionalities

### PR DESCRIPTION
**Pull request by sboeuf**
This pull request finishes the series by enabling snapshot/restore functionality through the VM. It has limited scope for now as it does not support virtio and VFIO devices, but that's the first step in supporting snapshotting and restoring a VM.

### CODE REVIEW

**Comment by sboeuf** [comment]
/!\ This PR depends on https://github.com/cloud-hypervisor/cloud-hypervisor/pull/995 /!\
Do NOT merge before https://github.com/cloud-hypervisor/cloud-hypervisor/pull/995 gets merged.

**Comment by sboeuf** [comment]
One more thing, this PR needs an integration test that I will add tomorrow. But it is ready to be reviewed though.

File: vmm/src/cpu.rs [hunk 1]
```diff
@@ -823,6 +875,107 @@ impl CpuManager {
...
+        );
+
+        // On hot plug calls into this function entry_point is None. It is for
+        // those hotplug CPU additions that we need to set the inserting flag.
+        self.vcpu_states[usize::from(cpu_id)].handle = handle;
+        self.vcpu_states[usize::from(cpu_id)].inserting = false;
+
+        Ok(())
+    }
+
```
Comment (sameo) [inline vmm/src/cpu.rs:979]: I have to blame myself for that one, and this was part of my TODO list, but there's a lot of common code between `restore_vcpu()` and the main `activate_vcpus()` loop. We have to factorize the vcpu thread creation code at least.

**Comment by sboeuf** [comment]
>One more thing, this PR needs an integration test that I will add tomorrow. But it is ready to be reviewed though.

Ok so we decided we won't add an integration for now, as we want something simple relying on ssh. Once the snapshot/restore feature will be supported for virtio-net, we will be able to add this test.

File: vmm/src/cpu.rs (hunk 1 above)
Comment (sboeuf) [reply vmm/src/cpu.rs:979]: Ok I'm gonna look at this!

Comment (sboeuf) [reply vmm/src/cpu.rs]: Done!

File: vmm/src/cpu.rs [hunk 2]
```diff
@@ -545,6 +739,153 @@ impl CpuManager {
...
+                .expect("Failed to restore vCPU");
+        } else {
+            let vm_memory = self.vm_memory.clone();
+
+            vcpu.lock()
+                .unwrap()
+                .configure(entry_point, &vm_memory, self.cpuid.clone())
+                .expect("Failed to configure vCPU");
+        }
+
```
Comment (sameo) [inline vmm/src/cpu.rs:829]: Nice, thanks for that.

File: vmm/src/device_manager.rs [hunk 3]
```diff
@@ -591,6 +591,7 @@ impl DeviceManager {
```
Comment (rbradford) [inline vmm/src/device_manager.rs:593]: This commit message is too long:

`vmm: device_manger: Make serial and ioapic devices migratable`

File: vmm/Cargo.toml [hunk 4]
```diff
@@ -19,7 +19,7 @@ anyhow = "1.0"
 arch = { path = "../arch" }
```
Comment (rbradford) [inline vmm/Cargo.toml:19]: This commit message could do with a "why".

File: vmm/src/cpu.rs [hunk 5]
```diff
@@ -265,17 +265,17 @@ impl Vcpu {
         mmio_bus: Arc<devices::Bus>,
```
Comment (rbradford) [inline vmm/src/cpu.rs:341]: This commit message needs a "why"

File: vmm/src/cpu.rs [hunk 6]
```diff
@@ -384,6 +455,65 @@ impl Vcpu {
             ts.as_micros()
         );
     }
+
+    #[allow(unused)]
```
Comment (rbradford) [inline vmm/src/cpu.rs]: I don't see these being removed in a later commit.

Comment (sboeuf) [reply vmm/src/cpu.rs]: Yes it's removed here: https://github.com/cloud-hypervisor/cloud-hypervisor/pull/998/commits/a7c6560347f302fcab7826900c7e086f570f4b1f#diff-39805f31fede77dcaa77b9055d65e821L459

File: vmm/src/cpu.rs (hunk 5 above)
Comment (sboeuf) [reply vmm/src/cpu.rs:341]: I will update it!

File: vmm/Cargo.toml (hunk 4 above)
Comment (sboeuf) [reply vmm/Cargo.toml:19]: I will update it!

File: vmm/src/device_manager.rs (hunk 3 above)
Comment (sboeuf) [reply vmm/src/device_manager.rs:593]: Ok!

**Comment by sboeuf** [comment]
@rbradford @sameo Ok I think I addressed your comments by expanding every commit message. Also, I implemented the copy of the memory region (filling the regions with the file content) instead of having the backing file.
As a follow up PR, I will look at supporting the VM template use case we've been talking about.

File: vmm/src/memory_manager.rs [hunk 7]
```diff
@@ -350,6 +353,66 @@ impl MemoryManager {
...
+                        )
+                        .map_err(|e| Error::Restore(MigratableError::MigrateReceive(e.into())))?;
+
+                    Ok(())
+                })?;
+        }
+
+        Ok(memory_manager)
+    }
+
```
Comment (sameo) [inline vmm/src/memory_manager.rs:415]: That looks simpler indeed, it's good.
Thanks for the refactoring, LGTM.

File: vmm/src/memory_manager.rs [hunk 8]
```diff
@@ -342,6 +353,66 @@ impl MemoryManager {
...
+                    }
+                };
+
+            memory_manager
+                .lock()
+                .unwrap()
+                .guest_memory
+                .memory()
+                .with_regions(|index, region| {
+                    let memory_region_path = PathBuf::from(format!(
```
Comment (rbradford) [inline vmm/src/memory_manager.rs:389]: Should really use the path functions for extending (does the right thing with extra/missing slashes). `PathBuf::push()` is what you want.

Comment (sboeuf) [reply vmm/src/memory_manager.rs:389]: Ok I'll update :)

Comment (sboeuf) [reply vmm/src/memory_manager.rs]: Done!

**Pull request merged**